File: problems/1001-1100/1095_find-in-mountain-array/provided/go/oracle.go
package main

// Problem-provided oracle (MountainArray), Go side. Compiled beside
// every submission by the judge; never editable in the editor.
// Constructed from the case state: the sequence's values as generic
// values, then the query budget.
type MountainArray struct {
	values []int
	budget int64
}

// NewMountainArray builds the oracle from the case's construction
// values (one generic slice of ints) and the query budget.
func NewMountainArray(construction []any, budget int64) *MountainArray {
	items, ok := construction[0].([]any)
	if !ok {
		panic("MountainArray values must be an array")
	}
	values := make([]int, 0, len(items))
	for _, item := range items {
		value, ok := item.(int64)
		if !ok {
			panic("MountainArray values must be integers")
		}
		values = append(values, int(value))
	}
	return &MountainArray{values: values, budget: budget}
}

// Get returns the value at index, spending one unit of query budget.
func (reader *MountainArray) Get(index int) int {
	if reader.budget <= 0 {
		panic("MountainArray query budget exhausted")
	}
	reader.budget--
	if index < 0 || index >= len(reader.values) {
		panic("MountainArray index out of range")
	}
	return reader.values[index]
}

// Length returns the sequence's length, free of charge.
func (reader *MountainArray) Length() int {
	return len(reader.values)
}

File: problems/1201-1300/1274_number-of-ships-in-a-rectangle/provided/rust/oracle.rs
// Problem-provided oracle (Sea), Rust side. Assembled into every
// submission's crate by the judge; never editable in the editor.
// Constructed from the case state: the hidden ship points as generic
// values, then the query budget.
#[allow(dead_code)]
pub struct Sea {
    ships: Vec<[i32; 2]>,
    budget: i64,
}

impl Sea {
    pub fn new(construction: &[OjValue], budget: i64) -> Self {
        let points = match construction.first() {
            Some(OjValue::Array(points)) => points.clone(),
            _ => panic!("Sea ship data must be an array"),
        };
        let mut ships = Vec::with_capacity(points.len());
        for point in points {
            let pair = match &point {
                OjValue::Array(pair) if pair.len() == 2 => pair,
                _ => panic!("Sea ship data must hold point pairs"),
            };
            let x = match pair[0] {
                OjValue::Int(v) => v as i32,
                _ => panic!("Sea ship points must be integers"),
            };
            let y = match pair[1] {
                OjValue::Int(v) => v as i32,
                _ => panic!("Sea ship points must be integers"),
            };
            ships.push([x, y]);
        }
        Sea { ships, budget }
    }

    pub fn has_ships(&mut self, top_right: &[i32], bottom_left: &[i32]) -> bool {
        if self.budget <= 0 {
            panic!("Sea query budget exhausted");
        }
        self.budget -= 1;
        self.ships.iter().any(|ship| {
            ship[0] >= bottom_left[0] && ship[0] <= top_right[0] && ship[1] >= bottom_left[1] && ship[1] <= top_right[1]
        })
    }
}
